### PR TITLE
Replacement over stale PR: Fix bytes mut asymptotics

### LINKED ISSUE #524: Bad “optimization” of `BytesMut` re-allocations
Reading 

https://github.com/tokio-rs/bytes/blob/68afb40df9ea2e61a93d0e26f6f8a9f231979452/src/bytes_mut.rs#L577-L579

I felt like this can completely kill the asymptotics of `BytesMut`. And sure enough, e.g. this example:

on a full `BytesMut`, a million bytes, doing 10'000 `get_u8` and `put_u8` operations moves the entire contents of the `BytesMut` 10'000 times, resulting in a run-time (in the playground) of up to 1 second, instead of less than a millisecond.

```rs
use bytes::{BytesMut, BufMut, Buf};

use std::time;

fn main() {
    let mut bytes = BytesMut::with_capacity(1024);
    bytes.put_bytes(0, 1_000_000);
    assert_eq!(bytes.capacity(), 1_000_000);
    let start = time::Instant::now();
    for _ in 0..10_000 {
        bytes.get_u8();
        bytes.put_u8(0);
    }
    println!("{:>9.3?} - 10'000 × get+put (each moving all the data)", start.elapsed());
    let start = time::Instant::now();
    bytes.put_u8(0);
    println!("{:>9.3?} - one put (re-allocating)", start.elapsed());
    let start = time::Instant::now();
    for _ in 0..10_000 {
        bytes.get_u8();
        bytes.put_u8(0);
    }
    println!("{:>9.3?} - 10'000 × get+put (no allocations or moves)", start.elapsed());
}
```
```plain
633.147ms - 10'000 × get+put (each moving all the data)
 71.792µs - one put (re-allocating)
223.719µs - 10'000 × get+put (no allocations or moves)
```
([playground](https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=61594b1797a3faab78e715101463f29b))

The optimization here is bad, since usually automatic re-allocations follow the principle that the capacity has to grow by a constant factor each time, whereas moving the data might only increase the capacity by a tiny bit. This isn't a re-allocation; it just copies all the data, but really the actual “allocation” part of a re-allocation is cheap, it always is the copying that it's all about.

Vec (currently) uses a [factor 2](https://doc.rust-lang.org/1.57.0/src/alloc/raw_vec.rs.html#440), so I suggest that this “optimization” should really only ever trigger if the gained capacity, i.e. `self.capacity() - self.len() + off`, increases the previous capacity, `self.capacity()` at least by a factor of two, i.e. if
* (current condition still relevant): `self.capacity() - self.len() + off >= additional`, **and also**
* (new condition): `off >= self.capacity()`

Interesting consequence: This also has the effect that `copy_nonoverlapping` can be used.

### PR DESCRIPTION
**Pull request by NobodyXu**
Replacement for #525 as it has been stale for really long.

Fixes #524 
Closes #525

### CODE REVIEW

**Comment by NobodyXu** [comment]
@Darksonn Request for review please.

I have rebased #525 against master, this fixed the sanitizer and stable/nightly CI.

I also fixed the minrust CI by removing use of `offset_from` and fixed the "must-use" error in `{bytes, bytes_mut}::release_shared` by manually `drop`ping `Box::from_raw`.

File: src/bytes_mut.rs [hunk 1]
```diff
@@ -636,11 +663,25 @@ impl BytesMut {
...
+                let v_capacity = v.capacity();
+                let ptr = v.as_mut_ptr();
+
+                // The following line is equivalent to:
+                //
+                //     self.ptr.as_ptr().offset_from(ptr) as usize;
+                //
+                // But due to min rust is 1.39 and it is only stablised
+                // in 1.47, we cannot use it.
+                let offset = self.ptr.as_ptr() as usize - ptr as usize;
```
Comment (Darksonn) [inline src/bytes_mut.rs:675]: In bytes, we usually put this kind of functionality in helper functions. Can you move this into an `offset_from` helper fn? You can put it next to the `invalid_ptr` helper fn.

**Review by Darksonn — COMMENTED** [review]
This looks reasonable. The documentation is well written.

I have one nit:

File: src/bytes_mut.rs (hunk 1 above)
Comment (NobodyXu) [reply src/bytes_mut.rs:675]: Sure

Comment (NobodyXu) [reply src/bytes_mut.rs]: Done!

**Comment by NobodyXu** [comment]
> The documentation is well written.

They are actually written by @steffahn , I just rebased them against master and fixed the CI.

**Review by Darksonn — APPROVED** [review]
Thanks @NobodyXu and @steffahn!

**Comment by NobodyXu** [comment]
@Darksonn Can we make a new release? It has been a long time since the last release.

**Comment by Darksonn** [comment]
Yeah, it looks like it's time for a release.

**Comment by Darksonn** [comment]
> code seemingly relying on (maybe(?) questionable) things such as that the uninitialized portion of a Vec gets preserved during a resize

The standard library `Vec` certainly doesn't guarantee that the uninitialized portion is preserved.

**Comment by steffahn** [comment]
> I fixed them and it just works.

Oh, well, that's nice I guess 😅. I wasn't really thinking there were any problems in my change of the code anyways (as also indicated by the fact that, at the time, problems only started to appear after some changes on master).

**Comment by steffahn** [comment]
> The standard library `Vec` certainly doesn't guarantee that the uninitialized portion is preserved.

Alright. I'll try to find where in `bytes` I thought I saw code that seemed to rely on this, and check if it still exists and if yes, open an issue 😇

**Comment by Darksonn** [comment]
If you look at the release PR I just posted, you'll see quite a few PRs related to fixing miri. It was probably those that fixed most of the CI issues on your PR.

> Alright. I'll try to find where in bytes I thought I saw code that seemed to rely on this, and check if it still exists and if yes, open an issue 

Thanks!

**Comment by steffahn** [comment]
> Alright. I'll try to find where in `bytes` I thought I saw code that seemed to rely on this, and check if it still exists and if yes, open an issue 😇

Haven’t looked back into the code yet, instead I tried a more systematic approach of utilizing miri as a first step, with a modified standard library that avoids copying any not-considered-initialized portion of the `Vec`
```diff
diff --git a/library/alloc/src/raw_vec.rs b/library/alloc/src/raw_vec.rs
index b0f4529abdf..b32afee9e43 100644
--- a/library/alloc/src/raw_vec.rs
+++ b/library/alloc/src/raw_vec.rs
@@ -397,7 +397,12 @@ fn grow_amortized(&mut self, len: usize, additional: usize) -> Result<(), TryRes
         let new_layout = Layout::array::<T>(cap);
 
         // `finish_grow` is non-generic over `T`.
-        let ptr = finish_grow(new_layout, self.current_memory(), &mut self.alloc)?;
+        let ptr = finish_grow(
+            len * mem::size_of::<T>(),
+            new_layout,
+            self.current_memory(),
+            &mut self.alloc,
+        )?;
         self.set_ptr_and_cap(ptr, cap);
         Ok(())
     }
@@ -416,7 +421,12 @@ fn grow_exact(&mut self, len: usize, additional: usize) -> Result<(), TryReserve
         let new_layout = Layout::array::<T>(cap);
 
         // `finish_grow` is non-generic over `T`.
-        let ptr = finish_grow(new_layout, self.current_memory(), &mut self.alloc)?;
+        let ptr = finish_grow(
+            len * mem::size_of::<T>(),
+            new_layout,
+            self.current_memory(),
+            &mut self.alloc,
+        )?;
         self.set_ptr_and_cap(ptr, cap);
         Ok(())
     }
@@ -446,6 +456,7 @@ fn shrink(&mut self, cap: usize) -> Result<(), TryReserveError> {
 // much smaller than the number of `T` types.)
 #[inline(never)]
 fn finish_grow<A>(
+    len_u8: usize,
     new_layout: Result<Layout, LayoutError>,
     current_memory: Option<(NonNull<u8>, Layout)>,
     alloc: &mut A,
@@ -463,7 +474,16 @@ fn finish_grow<A>(
         unsafe {
             // The allocator checks for alignment equality
             intrinsics::assume(old_layout.align() == new_layout.align());
-            alloc.grow(ptr, old_layout, new_layout)
+            let new = alloc.allocate(new_layout);
+            if let Ok(new_ptr) = new {
+                ptr::copy_nonoverlapping(
+                    ptr.as_ptr() as *const MaybeUninit<u8>,
+                    new_ptr.as_mut_ptr() as *mut MaybeUninit<u8>,
+                    len_u8,
+                );
+                alloc.deallocate(ptr, old_layout);
+            }
+            new
         }
     } else {
         alloc.allocate(new_layout)
```
(applied to 866d5c621c9b87f82a5aa53c3dfc1e477181f33b)

and provided I didn’t mess up that “growing” code, there ***is*** still a problem
(using `miri 0.1.0 (b938529 2022-07-25)` that comes with `rustc 1.64.0-nightly (4493a0f47 2022-08-02))
```
rm -rf ~/.cache/miri; MIRI_LIB_SRC=PATH/TO/CHECKOUT/OF/rust/library cargo +nightly miri test
```
```
test reserve_shared_reuse ... error: Undefined Behavior: reading memory at alloc456687[0xd..0x16], but memory is uninitialized at [0xd..0x16], and this operation requires initialized memory
   --> /home/frank/forks/rust/library/core/src/array/equality.rs:152:13
    |
152 |             crate::intrinsics::raw_eq(a, b)
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ reading memory at alloc456687[0xd..0x16], but memory is uninitialized at [0xd..0x16], and this operation requires initialized memory
    |
    = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
    = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
    = note: backtrace:
    = note: inside `<u8 as std::array::equality::SpecArrayEq<u8, 9>>::spec_eq` at /home/frank/forks/rust/library/core/src/array/equality.rs:152:13
    = note: inside `std::array::equality::<impl std::cmp::PartialEq for [u8; 9]>::eq` at /home/frank/forks/rust/library/core/src/array/equality.rs:12:9
    = note: inside `std::array::equality::<impl std::cmp::PartialEq<[u8; 9]> for [u8]>::eq` at /home/frank/forks/rust/library/core/src/array/equality.rs:52:22
    = note: inside `std::cmp::impls::<impl std::cmp::PartialEq<&[u8; 9]> for &[u8]>::eq` at /home/frank/forks/rust/library/core/src/cmp.rs:1513:13
note: inside `reserve_shared_reuse` at /home/frank/forks/rust/library/core/src/macros/mod.rs:40:21
   --> tests/test_bytes.rs:573:5
    |
573 |     assert_eq!(&*bytes, b"!123ex123");
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside closure at tests/test_bytes.rs:562:1
   --> tests/test_bytes.rs:562:1
    |
561 |   #[test]
    |   ------- in this procedural macro expansion
562 | / fn reserve_shared_reuse() {
563 | |     let mut bytes = BytesMut::with_capacity(1000);
564 | |     bytes.put_slice(b"Hello, World!");
565 | |     drop(bytes.split());
...   |
574 | |     assert_eq!(bytes.capacity(), 2009);
575 | | }
    | |_^
    = note: this error originates in the macro `assert_eq` which comes from the expansion of the attribute macro `test` (in Nightly builds, run with -Z macro-backtrace for more info)

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace
```

**Pull request merged**